service/src/main/java/com/epam/esm/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,246 @@
+        List<TagDTO> tagDTOList = giftCertificateDTO.getTags();
+        if (tagDTOList.size() != 0) {
+            tagDTOList.forEach(tagDTO -> {
+                Optional<Tag> tagByName = tagRepository.findByName(tagDTO.getName());
+                if (!tagByName.isPresent()) {
+                    Long tagId = tagRepository.create(tagConverter.toEntity(tagDTO));
+                    tagRepository.createConnectionBetweenTagAndGiftCertificate(tagId, id);
+                } else {
+                    tagRepository.createConnectionBetweenTagAndGiftCertificate(tagByName.get().getId(), id);
+                }
+            });
+        }
+        return id;
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<GiftCertificateDTO> findAll() {
+        List<GiftCertificate> giftCertificates = giftCertificateRepository.findAll();
+        List<GiftCertificateDTO> giftCertificateDTOS = giftCertificates.stream()
+                .map(giftCertificateConverter::toDTO).collect(Collectors.toList());
+        giftCertificateDTOS.forEach(giftCertificateDTO -> {
+            List<Tag> tags = tagRepository.findListTagsByCertificateId(giftCertificateDTO.getId());
+            List<TagDTO> tagDTOList = tags.stream().map(tagConverter::toDTO).collect(Collectors.toList());
+            giftCertificateDTO.setTags(tagDTOList);
+        });
+        return giftCertificateDTOS;
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public GiftCertificateDTO findByName(String name) throws ServiceException {
+        GiftCertificateDTO giftCertificateDTO = null;
+        Optional<GiftCertificate> giftCertificate = giftCertificateRepository.findByName(name);
+        if (giftCertificate.isPresent()) {
+            List<Tag> tags = tagRepository.findListTagsByCertificateId(giftCertificate.get().getId());
+            giftCertificateDTO = giftCertificateConverter.toDTO(giftCertificate.get());
+            List<TagDTO> tagDTOList = tags.stream().map(tagConverter::toDTO).collect(Collectors.toList());
+            giftCertificateDTO.setTags(tagDTOList);
+        }
+        return giftCertificateDTO;
+    }
+
+    @Override
+    @Transactional
+    public boolean update(GiftCertificateDTO giftCertificateDTO) {
+        boolean result;
+        GiftCertificate giftCertificate = giftCertificateRepository.findById(giftCertificateDTO.getId())
+                .orElseThrow(() -> new ServiceException("Certificate not found with id " + giftCertificateDTO.getId()));
+
+        Optional<GiftCertificate> giftCertificateByName = giftCertificateRepository.findByName(giftCertificateDTO.getName());
+        if (giftCertificateByName.isPresent() && giftCertificateByName.get().getId() != giftCertificateDTO.getId()) {
+            throw new ServiceException("Certificate found with name " + giftCertificateDTO.getName());
+        }
+        GiftCertificate updateGiftCertificate = new GiftCertificate();
```

Comment:
why not use giftCertificateConverter here `GiftCertificate giftCertificate = giftCertificateConverter.toEntity(giftCertificateDTO);` ?